File: Algoritmos de ordenamiento/py/11 Counting.py
def counting_sort(array):
    size = len(array)

    min_value = min(array)
    max_value = max(array)
    range_value = max_value - min_value + 1
    count = [0] * range_value

    for num in array:
        count[num - min_value] += 1

    index = 0
    for i in range(range_value):
        while count[i] > 0:
            array[index] = i + min_value
            count[i] -= 1
            index += 1

array = [2, 3, 1, 923, -23, 9128, 0, 10, 20, 12]

print("\n== Estos son tus valores == \n")
print(*array, sep="\t")

print("\n\n== Ordenando... ==\n")

counting_sort(array)

for num in array:
    print(num, end="\t")
print()

print("\n== Estos son tus valores ordenados ==\n")
print(*array, sep="\t")

print("\n")

input("Presiona Enter para continuar...")

File: Algoritmos de ordenamiento/cpp/11 Counting.cpp
#include <iostream>
#include <vector>
using namespace std;

int main() {
    system("title Algoritmos de ordenamientos de datos en C++ / Countingsort && cls");

    int cache = 0;

    int array[10] = {2, 3, 1, 923, -23, 9128, 0, 10, 20, 12};
    size_t size = sizeof(array) / sizeof(array[0]);

    cout << "\n== Estos son tus valores == \n\n";
    for (int i = 0; i < size; i++) {
        cout << array[i] << "    ";
    }

    cout << "\n\n== Ordenando... ==\n\n";

    int minValue = array[0];
    int maxValue = array[0];
    for (int i = 1; i < size; i++) {
        if (array[i] < minValue) {
            minValue = array[i];
        }
        if (array[i] > maxValue) {
            maxValue = array[i];
        }
    }

    int range = maxValue - minValue + 1;
    vector<int> count(range, 0);

    for (int i = 0; i < size; i++) {
        count[array[i] - minValue]++;

        for (int i = 0; i < size; i++) {
            cout << array[i] << "    ";
        }
        cout << endl;
    }

    int index = 0;
    for (int i = 0; i < range; i++) {
        while (count[i] > 0) {
            array[index++] = i + minValue;
            count[i]--;

            for (int i = 0; i < size; i++) {
                cout << array[i] << "    ";
            }
            cout << endl;
        }
    }

    cout << "\n== Estos son tus valores ordenados ==\n\n";
    for (int i = 0; i < size; i++) {
        cout << array[i] << "    ";
    }

    cout << endl << endl;

    system("pause");
    system("cls");
}
